File: packages/esm-format-go/pkg/esm/canonicalize.go
package esm

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"math"
	"sort"
	"strconv"
	"strings"
)

// Error codes per discretization RFC §5.4.6 / §5.4.7.
var (
	ErrCanonicalNonFinite = errors.New("E_CANONICAL_NONFINITE")
	ErrCanonicalDivByZero = errors.New("E_CANONICAL_DIVBY_ZERO")
)

// Canonicalize applies RFC §5.4 canonical form to an expression tree.
//
// Returns ErrCanonicalNonFinite if the tree contains NaN or ±Inf, and
// ErrCanonicalDivByZero for 0/0. Input is not mutated; callers receive a new
// tree built from the same leaf values.
func Canonicalize(expr Expression) (Expression, error) {
	switch e := expr.(type) {
	case nil:
		return nil, fmt.Errorf("nil expression")
	case int:
		return int64(e), nil
	case int32:
		return int64(e), nil
	case int64:
		return e, nil
	case float32:
		return canonFloat(float64(e))
	case float64:
		return canonFloat(e)
	case string:
		return e, nil
	case ExprNode:
		return canonOp(e)
	case *ExprNode:
		return canonOp(*e)
	default:
		return nil, fmt.Errorf("unknown expression type: %T", expr)
	}
}

func canonFloat(f float64) (Expression, error) {
	if math.IsNaN(f) || math.IsInf(f, 0) {
		return nil, ErrCanonicalNonFinite
	}
	return f, nil
}

func canonOp(node ExprNode) (Expression, error) {
	// Step 1: recursively canonicalize every child.
	newArgs := make([]interface{}, len(node.Args))
	for i, a := range node.Args {
		ca, err := Canonicalize(a)
		if err != nil {
			return nil, err
		}
		newArgs[i] = ca
	}
	work := ExprNode{
		Op:        node.Op,
		Args:      newArgs,
		Wrt:       node.Wrt,
		Dim:       node.Dim,
		HandlerID: node.HandlerID,
	}

	// Step 2: operator-specific rewrites.
	switch work.Op {
	case "+":
		return canonAdd(work)
	case "*":
		return canonMul(work)
	case "-":
		return canonSub(work)
	case "/":
		return canonDiv(work)
	case "neg":
		return canonNeg(work)
	default:
		return work, nil
	}
}

// canonAdd: flatten, eliminate zeros (type-preserving), order, collapse singletons.
func canonAdd(node ExprNode) (Expression, error) {
	flat := flattenSameOp(node.Args, "+")
	others, hadIntZero, hadFloatZero := partitionIdentity(flat, 0)
	_ = hadIntZero
	// Float-zero is only safe to drop when all survivors are float literals
	// (otherwise we lose the float-promotion hint). If unsafe, keep one 0.0.
	if hadFloatZero && !allFloatLiterals(others) {
		others = append(others, 0.0)
	}
	if len(others) == 0 {
		if hadFloatZero {
			return 0.0, nil
		}
		return int64(0), nil
	}
	if len(others) == 1 {
		return others[0], nil
	}
	sortArgs(others)
	return ExprNode{Op: "+", Args: others}, nil
}

// canonMul: flatten, zero-annihilation, identity elim (type-preserving), order.
func canonMul(node ExprNode) (Expression, error) {
	flat := flattenSameOp(node.Args, "*")
	// Zero annihilation (§5.4.4): preserve the numeric type of the zero.
	for _, a := range flat {
		if exprIsZeroInt(a) {
			return int64(0), nil
		}
		if exprIsZeroFloat(a) {
			if f, ok := a.(float64); ok {
				return f * 0.0, nil // preserves -0.0 signbit
			}
			return 0.0, nil
		}
	}
	others, hadIntOne, hadFloatOne := partitionIdentity(flat, 1)
	_ = hadIntOne
	if hadFloatOne && !allFloatLiterals(others) {
		others = append(others, 1.0)
	}
	if len(others) == 0 {
		if hadFloatOne {
			return 1.0, nil
		}
		return int64(1), nil
	}
	if len(others) == 1 {
		return others[0], nil
	}
	sortArgs(others)
	return ExprNode{Op: "*", Args: others}, nil
}

// partitionIdentity splits args into (non-identity, hadIntIdentity, hadFloatIdentity)
// where identityValue is 0 (for +) or 1 (for *).
func partitionIdentity(args []interface{}, identityValue int64) (others []interface{}, hadInt, hadFloat bool) {
	others = make([]interface{}, 0, len(args))
	for _, a := range args {
		switch v := a.(type) {
		case int64:
			if v == identityValue {
				hadInt = true
				continue
			}
		case float64:
			if v == float64(identityValue) {
				hadFloat = true
				continue
			}
		}
		others = append(others, a)
	}
	return others, hadInt, hadFloat
}

func allFloatLiterals(args []interface{}) bool {
	if len(args) == 0 {
		return false
	}
	for _, a := range args {
		if !exprIsFloat(a) {
			return false
		}
	}
	return true
}

// canonSub: kept as distinct op, preserve arg order, apply identity rules.
// Convert {-, 0, x} to {neg, x} when x is not a literal.
func canonSub(node ExprNode) (Expression, error) {
	if len(node.Args) == 1 {
		// Unary form is typically spelled `neg` on the wire, but tolerate.
		return canonNeg(ExprNode{Op: "neg", Args: node.Args})
	}
	if len(node.Args) == 2 {
		a, b := node.Args[0], node.Args[1]
		// -(0, x) -> neg(x) (with type-preserving: -(0, x_literal) folds to negated literal)
		if exprIsZeroAny(a) {
			return canonNeg(ExprNode{Op: "neg", Args: []interface{}{b}})
		}
		// -(x, 0) -> x, type-preserving: if 0 is float and x is int literal, promote.
		if exprIsZeroAny(b) {
			if exprIsZeroFloat(b) && exprIsIntLiteral(a) {
				return float64(a.(int64)), nil
			}
			return a, nil
		}
	}
	return node, nil
}

// canonDiv: kept as distinct op, preserve order, identity rules, 0/0 error.
func canonDiv(node ExprNode) (Expression, error) {
	if len(node.Args) != 2 {
		return node, nil
	}
	a, b := node.Args[0], node.Args[1]
	if exprIsZeroAny(a) && exprIsZeroAny(b) {
		return nil, ErrCanonicalDivByZero
	}
	// /(x, 1) -> x, with type-preserving.
	if exprIsOneAny(b) {
		if exprIsOneFloat(b) && exprIsIntLiteral(a) {
			return float64(a.(int64)), nil
		}
		return a, nil
	}
	// /(0, x) -> 0 when x != 0 (literally-zero test only; structural x is unknown).
	if exprIsZeroAny(a) && !exprIsZeroAny(b) {
		if exprIsZeroFloat(a) {
			return 0.0, nil
		}
		return int64(0), nil
	}
	return node, nil
}

// canonNeg: neg(neg(x))->x, neg(literal)->negated literal, neg(0)->0.
func canonNeg(node ExprNode) (Expression, error) {
	if len(node.Args) != 1 {
		return node, nil
	}
	x := node.Args[0]
	switch v := x.(type) {
	case int64:
		return -v, nil
	case float64:
		return -v, nil
	case ExprNode:
		if v.Op == "neg" && len(v.Args) == 1 {
			return v.Args[0], nil
		}
	case *ExprNode:
		if v.Op == "neg" && len(v.Args) == 1 {
			return v.Args[0], nil
		}
	}
	return ExprNode{Op: "neg", Args: []interface{}{x}}, nil
}

// flattenSameOp inlines nested same-op children.
func flattenSameOp(args []interface{}, op string) []interface{} {
	out := make([]interface{}, 0, len(args))
	for _, a := range args {
		switch v := a.(type) {
		case ExprNode:
			if v.Op == op {
				out = append(out, v.Args...)
				continue
			}
		case *ExprNode:
			if v.Op == op {
				out = append(out, v.Args...)
				continue
			}
		}
		out = append(out, a)
	}
	return out
}

func exprIsFloat(a interface{}) bool {
	_, ok := a.(float64)
	return ok
}

func exprIsIntLiteral(a interface{}) bool {
	_, ok := a.(int64)
	return ok
}

func exprIsZeroAny(a interface{}) bool {
	switch v := a.(type) {
	case int64:
		return v == 0
	case float64:
		return v == 0.0
	}
	return false
}

func exprIsZeroInt(a interface{}) bool {
	v, ok := a.(int64)
	return ok && v == 0
}

func exprIsZeroFloat(a interface{}) bool {
	v, ok := a.(float64)
	return ok && v == 0.0
}

func exprIsOneAny(a interface{}) bool {
	switch v := a.(type) {
	case int64:
		return v == 1
	case float64:
		return v == 1.0
	}
	return false
}

func exprIsOneFloat(a interface{}) bool {
	v, ok := a.(float64)
	return ok && v == 1.0
}

// sortArgs sorts args in place per §5.4.2.
//
//  1. Numeric literals first, ascending value, int-before-float at equal magnitude.
//  2. Bare strings lexicographically.
//  3. Non-leaf nodes by canonical JSON byte compare.
func sortArgs(args []interface{}) {
	// Memoize the canonical JSON for non-leaf nodes to avoid quadratic serialization.
	jsonCache := make(map[int]string)
	getJSON := func(idx int, a interface{}) string {
		if s, ok := jsonCache[idx]; ok {
			return s
		}
		s, _ := emitCanonicalJSON(a)
		jsonCache[idx] = s
		return s
	}
	// Build an index-preserving slice.
	n := len(args)
	idx := make([]int, n)
	for i := range idx {
		idx[i] = i
	}
	sort.SliceStable(idx, func(i, j int) bool {
		return argLess(args[idx[i]], args[idx[j]], idx[i], idx[j], getJSON)
	})
	sorted := make([]interface{}, n)
	for i, k := range idx {
		sorted[i] = args[k]
	}
	copy(args, sorted)
}

func argTier(a interface{}) int {
	switch a.(type) {
	case int64, float64:
		return 0
	case string:
		return 1
	case ExprNode, *ExprNode:
		return 2
	}
	return 3
}

func argLess(a, b interface{}, ia, ib int, getJSON func(int, interface{}) string) bool {
	ta, tb := argTier(a), argTier(b)
	if ta != tb {
		return ta < tb
	}
	switch ta {
	case 0:
		av, bv, af, bf := numericKey(a), numericKey(b), exprIsFloat(a), exprIsFloat(b)
		if av != bv {
			return av < bv
		}
		// At equal magnitude, int before float.
		return !af && bf
	case 1:
		return a.(string) < b.(string)
	case 2:
		return getJSON(ia, a) < getJSON(ib, b)
	}
	return false
}

func numericKey(a interface{}) float64 {
	switch v := a.(type) {
	case int64:
		return float64(v)
	case float64:
		return v
	}
	return 0
}

// CanonicalJSON emits the canonical on-wire JSON form of an expression per
// §5.4.6: keys sorted, no extraneous whitespace, shortest-round-trip float
// literals with trailing-`.0` disambiguation for integer-valued floats, and
// strict lowercase-e exponent notation without a leading `+`.
//
// The input is canonicalized first; pass an already-canonical tree for a
// no-op canonicalization pass.
func CanonicalJSON(expr Expression) ([]byte, error) {
	c, err := Canonicalize(expr)
	if err != nil {
		return nil, err
	}
	s, err := emitCanonicalJSON(c)
	if err != nil {
		return nil, err
	}
	return []byte(s), nil
}

func emitCanonicalJSON(a interface{}) (string, error) {
	switch v := a.(type) {
	case int64:
		return strconv.FormatInt(v, 10), nil
	case float64:
		if math.IsNaN(v) || math.IsInf(v, 0) {
			return "", ErrCanonicalNonFinite
		}
		return formatCanonicalFloat(v), nil
	case string:
		b, err := json.Marshal(v)
		if err != nil {
			return "", err
		}
		return string(b), nil
	case ExprNode:
		return emitExprNodeJSON(v)
	case *ExprNode:
		return emitExprNodeJSON(*v)
	case bool:
		if v {
			return "true", nil
		}
		return "false", nil
	case nil:
		return "null", nil
	}
	// Fall back to encoding/json for other types (kept for safety; not
	// expected in canonical ASTs).
	b, err := json.Marshal(a)
	if err != nil {
		return "", err
	}
	return string(b), nil
}

func emitExprNodeJSON(n ExprNode) (string, error) {
	// Collect key-value pairs then sort by key.
	kv := make([][2]string, 0, 4)
	// op
	opJSON, err := json.Marshal(n.Op)
	if err != nil {
		return "", err
	}
	kv = append(kv, [2]string{"op", string(opJSON)})
	// args
	argParts := make([]string, len(n.Args))
	for i, a := range n.Args {
		s, err := emitCanonicalJSON(a)
		if err != nil {
			return "", err
		}
		argParts[i] = s
	}
	kv = append(kv, [2]string{"args", "[" + strings.Join(argParts, ",") + "]"})
	// Optional fields (only emit when set).
	if n.Wrt != nil {
		b, _ := json.Marshal(*n.Wrt)
		kv = append(kv, [2]string{"wrt", string(b)})
	}
	if n.Dim != nil {
		b, _ := json.Marshal(*n.Dim)
		kv = append(kv, [2]string{"dim", string(b)})
	}
	if n.HandlerID != nil {
		b, _ := json.Marshal(*n.HandlerID)
		kv = append(kv, [2]string{"handler_id", string(b)})
	}
	sort.Slice(kv, func(i, j int) bool { return kv[i][0] < kv[j][0] })
	var buf bytes.Buffer
	buf.WriteByte('{')
	for i, p := range kv {
		if i > 0 {
			buf.WriteByte(',')
		}
		kb, _ := json.Marshal(p[0])
		buf.Write(kb)
		buf.WriteByte(':')
		buf.WriteString(p[1])
	}
	buf.WriteByte('}')
	return buf.String(), nil
}

// formatCanonicalFloat renders a finite float64 per §5.4.6: shortest
// round-trip decimal; plain decimal when 1e-6 <= |x| < 1e21 (with trailing
// `.0` added for integer-valued magnitudes); exponent notation with lowercase
// `e` and no leading `+` otherwise. Negative zero emits as `-0.0`.
func formatCanonicalFloat(f float64) string {
	if f == 0 {
		if math.Signbit(f) {
			return "-0.0"
		}
		return "0.0"
	}
	abs := math.Abs(f)
	useExp := abs < 1e-6 || abs >= 1e21
	if useExp {
		s := strconv.FormatFloat(f, 'e', -1, 64)
		// Go emits e.g. "1e+21", "3e-7"; spec requires no leading + on exp.
		if i := strings.IndexByte(s, 'e'); i != -1 {
			mant := s[:i]
			exp := s[i+1:]
			exp = strings.TrimPrefix(exp, "+")
			// Strip leading zeros on the exponent while preserving the sign:
			// Go emits e.g. "1e-07" for very small numbers on some platforms;
			// ECMAScript emits "1e-7".
			negExp := false
			if strings.HasPrefix(exp, "-") {
				negExp = true
				exp = exp[1:]
			}
			exp = strings.TrimLeft(exp, "0")
			if exp == "" {
				exp = "0"
			}
			if negExp {
				exp = "-" + exp
			}
			s = mant + "e" + exp
		}
		return s
	}
	// Plain decimal form.
	s := strconv.FormatFloat(f, 'f', -1, 64)
	if !strings.ContainsAny(s, ".eE") {
		s += ".0"
	}
	return s
}

File: packages/earthsci-toolkit/src/canonicalize.ts
/**
 * Canonical AST form per discretization RFC §5.4.
 *
 * Implements `canonicalize(expr)` and `canonicalJson(expr)` for the TypeScript
 * binding. Integer-vs-float AST-node distinction is carried by the
 * `NumericLiteral` tagged leaf (see `./numeric-literal.ts`); plain JS
 * `number` values are treated as untagged float-like literals for backward
 * compatibility with callers that have not migrated to `intLit` / `floatLit`.
 *
 * See `docs/rfcs/discretization.md` §5.4.1–§5.4.7 for the normative rules.
 */

import type { Expression } from './generated.js'
import {
  type NumericLiteral,
  intLit,
  floatLit,
  isFloatLit,
  isIntLit,
  isNumericLiteral,
} from './numeric-literal.js'

export class CanonicalizeError extends Error {
  /** Stable RFC §5.4.6 / §5.4.7 error code. */
  readonly code: string
  constructor(code: string, message?: string) {
    super(message ?? code)
    this.code = code
    this.name = 'CanonicalizeError'
  }
}

/** `E_CANONICAL_NONFINITE` — NaN or ±Inf (§5.4.6). */
export const E_CANONICAL_NONFINITE = 'E_CANONICAL_NONFINITE'
/** `E_CANONICAL_DIVBY_ZERO` — `/(0, 0)` (§5.4.7). */
export const E_CANONICAL_DIVBY_ZERO = 'E_CANONICAL_DIVBY_ZERO'

type Expr = Expression | NumericLiteral
type Node = { op: string; args: Expr[] } & Record<string, unknown>

function isNode(e: Expr): e is Node {
  return (
    typeof e === 'object' &&
    e !== null &&
    !isNumericLiteral(e) &&
    'op' in e &&
    'args' in e
  )
}

function isPlainNumber(e: Expr): e is number {
  return typeof e === 'number'
}

function isString(e: Expr): e is string {
  return typeof e === 'string'
}

/** True for plain JS numbers and tagged `NumericLiteral` leaves. */
function isNumericLeaf(e: Expr): boolean {
  return isPlainNumber(e) || isNumericLiteral(e)
}

/** Underlying numeric value for a plain number or `NumericLiteral`. */
function leafValue(e: Expr): number {
  if (isPlainNumber(e)) return e
  if (isNumericLiteral(e)) return e.value
  throw new TypeError('not a numeric leaf')
}

/**
 * A value counts as "float-like" for RFC §5.4.4 type-preserving rules when
 * it is either a plain JS number (TS has no native int tag) or an explicit
 * `floatLit`. Tagged `intLit` leaves and all non-numeric nodes are not
 * float-like.
 */
function isFloatLike(e: Expr): boolean {
  return isPlainNumber(e) || isFloatLit(e)
}

function isZeroValue(e: Expr): boolean {
  return isNumericLeaf(e) && leafValue(e) === 0
}

function isOneValue(e: Expr): boolean {
  return isNumericLeaf(e) && leafValue(e) === 1
}

/** Canonicalize an expression tree per RFC §5.4. Input is not mutated. */
export function canonicalize(expr: Expr): Expr {
  if (isNumericLiteral(expr)) {
    if (!Number.isFinite(expr.value)) {
      throw new CanonicalizeError(E_CANONICAL_NONFINITE)
    }
    return expr
  }
  if (isPlainNumber(expr)) {
    if (!Number.isFinite(expr)) {
      throw new CanonicalizeError(E_CANONICAL_NONFINITE)
    }
    return expr
  }
  if (isString(expr)) {
    return expr
  }
  if (isNode(expr)) {
    return canonOp(expr)
  }
  throw new TypeError(`unknown expression type: ${typeof expr}`)
}

/** Emit canonical on-wire JSON per §5.4.6 (sorted keys, no whitespace). */
export function canonicalJson(expr: Expr): string {
  return emitJson(canonicalize(expr))
}

function canonOp(node: Node): Expr {
  const newArgs: Expr[] = node.args.map((a) => canonicalize(a))
  const work: Node = { ...node, args: newArgs }
  switch (work.op) {
    case '+':
      return canonAdd(work)
    case '*':
      return canonMul(work)
    case '-':
      return canonSub(work)
    case '/':
      return canonDiv(work)
    case 'neg':
      return canonNeg(work)
    default:
      return work
  }
}

function canonAdd(node: Node): Expr {
  const flat = flattenSameOp(node.args, '+')
  const others: Expr[] = []
  let floatZeroSample: Expr | null = null
  let hadFloatZero = false
  let hadIntZero = false
  for (const a of flat) {
    if (isZeroValue(a)) {
      // Tagged float-zero is conditional (RFC §5.4.4 type-preserving).
      if (isFloatLit(a)) {
        if (floatZeroSample === null) floatZeroSample = a
        hadFloatZero = true
        continue
      }
      // Tagged int-zero and plain-number zeros drop unconditionally.
      if (isIntLit(a)) hadIntZero = true
      continue
    }
    others.push(a)
  }
  if (hadFloatZero && !others.every(isFloatLike)) {
    others.push(floatZeroSample as Expr)
    hadFloatZero = false // consumed — don't re-emit below
  }
  if (others.length === 0) {
    if (hadFloatZero) return floatZeroSample as Expr
    if (hadIntZero) return intLit(0)
    return 0
  }
  if (others.length === 1) return others[0]!
  sortArgs(others)
  return { ...node, op: '+', args: others }
}

function canonMul(node: Node): Expr {
  const flat = flattenSameOp(node.args, '*')
  // Zero annihilation (§5.4.4) — preserve the numeric type of the zero.
  for (const a of flat) {
    if (!isZeroValue(a)) continue
    if (isIntLit(a)) return intLit(0)
    if (isFloatLit(a)) return a // preserves signed zero
    // Plain number zero — preserve signed zero for -0.
    if (isPlainNumber(a) && Object.is(a, -0)) return -0
    return 0
  }
  const others: Expr[] = []
  let floatOneSample: Expr | null = null
  let hadFloatOne = false
  let hadIntOne = false
  for (const a of flat) {
    if (isOneValue(a)) {
      if (isFloatLit(a)) {
        if (floatOneSample === null) floatOneSample = a
        hadFloatOne = true
        continue
      }
      if (isIntLit(a)) hadIntOne = true
      continue
    }
    others.push(a)
  }
  if (hadFloatOne && !others.every(isFloatLike)) {
    others.push(floatOneSample as Expr)
    hadFloatOne = false
  }
  if (others.length === 0) {
    if (hadFloatOne) return floatOneSample as Expr
    if (hadIntOne) return intLit(1)
    return 1
  }
  if (others.length === 1) return others[0]!
  sortArgs(others)
  return { ...node, op: '*', args: others }
}

function canonSub(node: Node): Expr {
  if (node.args.length === 1) return canonNegValue(node.args[0]!)
  if (node.args.length === 2) {
    const a = node.args[0]!
    const b = node.args[1]!
    if (isZeroValue(a)) return canonNegValue(b)
    if (isZeroValue(b)) {
      // Type-preserving: -(intLit(x), floatLit(0)) → floatLit(x) to keep float promotion.
      if (isFloatLit(b) && isIntLit(a)) return floatLit(a.value)
      return a
    }
  }
  return node
}

function canonDiv(node: Node): Expr {
  if (node.args.length !== 2) return node
  const a = node.args[0]!
  const b = node.args[1]!
  if (isZeroValue(a) && isZeroValue(b)) {
    throw new CanonicalizeError(E_CANONICAL_DIVBY_ZERO)
  }
  if (isOneValue(b)) {
    if (isFloatLit(b) && isIntLit(a)) return floatLit(a.value)
    return a
  }
  if (isZeroValue(a)) {
    // /(0, x) — preserve zero type.
    if (isFloatLit(a)) return a
    if (isIntLit(a)) return intLit(0)
    return 0
  }
  return node
}

function canonNeg(node: Node): Expr {
  if (node.args.length !== 1) return node
  return canonNegValue(node.args[0]!)
}

function canonNegValue(arg: Expr): Expr {
  if (isIntLit(arg)) {
    // intLit cannot hold -0, so -0 value is impossible here.
    return arg.value === 0 ? arg : intLit(-arg.value)
  }
  if (isFloatLit(arg)) return floatLit(-arg.value)
  if (isPlainNumber(arg)) return -arg
  if (isNode(arg) && arg.op === 'neg' && arg.args.length === 1) return arg.args[0]!
  return { op: 'neg', args: [arg] }
}

function flattenSameOp(args: Expr[], op: string): Expr[] {
  const out: Expr[] = []
  for (const a of args) {
    if (isNode(a) && a.op === op) {
      out.push(...a.args)
    } else {
      out.push(a)
    }
  }
  return out
}

function argTier(e: Expr): number {
  if (isNumericLeaf(e)) return 0
  if (isString(e)) return 1
  if (isNode(e)) return 2
  return 3
}

function sortArgs(args: Expr[]): void {
  const cache = new Map<number, string>()
  const indices = args.map((_, i) => i)
  const getJson = (i: number, e: Expr): string => {
    let s = cache.get(i)
    if (s === undefined) {
      s = emitJson(e)
      cache.set(i, s)
    }
    return s
  }
  indices.sort((ia, ib) => compareExprs(args[ia]!, args[ib]!, ia, ib, getJson))
  const snap = indices.map((i) => args[i]!)
  for (let i = 0; i < args.length; i++) args[i] = snap[i]!
}

function compareExprs(
  a: Expr,
  b: Expr,
  ia: number,
  ib: number,
  getJson: (i: number, e: Expr) => string,
): number {
  const ta = argTier(a)
  const tb = argTier(b)
  if (ta !== tb) return ta - tb
  if (ta === 0) {
    const va = leafValue(a)
    const vb = leafValue(b)
    if (va !== vb) return va - vb
    // Equal magnitude: int before float (RFC §5.4.2 stable tiebreak).
    const af = isFloatLit(a) || isPlainNumber(a)
    const bf = isFloatLit(b) || isPlainNumber(b)
    if (af === bf) return 0
    return af ? 1 : -1
  }
  if (ta === 1) {
    return (a as string) < (b as string) ? -1 : (a as string) > (b as string) ? 1 : 0
  }
  const aj = getJson(ia, a)
  const bj = getJson(ib, b)
  return aj < bj ? -1 : aj > bj ? 1 : 0
}

function emitJson(e: Expr): string {
  if (isIntLit(e)) {
    // Value is guaranteed integer-valued and finite by intLit().
    const s = String(e.value)
    if (s.includes('.') || s.includes('e') || s.includes('E')) {
      throw new TypeError(`int NumericLiteral produced non-integer token ${s}`)
    }
    return s
  }
  if (isFloatLit(e)) {
    if (!Number.isFinite(e.value)) throw new CanonicalizeError(E_CANONICAL_NONFINITE)
    return formatCanonicalFloat(e.value)
  }
  if (isPlainNumber(e)) {
    if (!Number.isFinite(e)) throw new CanonicalizeError(E_CANONICAL_NONFINITE)
    return formatCanonicalFloat(e)
  }
  if (isString(e)) return JSON.stringify(e)
  if (isNode(e)) return emitNodeJson(e)
  if (e === null) return 'null'
  return JSON.stringify(e)
}

function emitNodeJson(n: Node): string {
  const entries: Array<[string, string]> = []
  entries.push(['op', JSON.stringify(n.op)])
  entries.push(['args', `[${n.args.map(emitJson).join(',')}]`])
  if ('wrt' in n && (n as Record<string, unknown>).wrt !== undefined) {
    entries.push(['wrt', JSON.stringify((n as Record<string, unknown>).wrt)])
  }
  if ('dim' in n && (n as Record<string, unknown>).dim !== undefined) {
    entries.push(['dim', JSON.stringify((n as Record<string, unknown>).dim)])
  }
  if ('handler_id' in n && (n as Record<string, unknown>).handler_id !== undefined) {
    entries.push(['handler_id', JSON.stringify((n as Record<string, unknown>).handler_id)])
  }
  entries.sort((x, y) => (x[0] < y[0] ? -1 : x[0] > y[0] ? 1 : 0))
  const body = entries.map(([k, v]) => `${JSON.stringify(k)}:${v}`).join(',')
  return `{${body}}`
}

/**
 * Format a finite `number` per RFC §5.4.6. Only handles float-typed
 * values: integer-typed `NumericLiteral` nodes are emitted as bare JSON
 * integers by {@link canonicalJson} directly.
 *
 * Unlike the convenience helper re-exported from `./numeric-literal`, this
 * version strips the leading `+` on exponent notation (RFC §5.4.6:
 * "no leading + on the exponent") so `1e25` emits as `1e25`, not `1e+25`.
 */
export function formatCanonicalFloat(f: number): string {
  if (!Number.isFinite(f)) {
    throw new CanonicalizeError(E_CANONICAL_NONFINITE)
  }
  if (f === 0) {
    return Object.is(f, -0) ? '-0.0' : '0.0'
  }
  const abs = Math.abs(f)
  const useExp = abs < 1e-6 || abs >= 1e21
  if (useExp) {
    return normalizeExponent(f.toString())
  }
  let s = f.toString()
  if (s.includes('e') || s.includes('E')) {
    s = expandToPlain(f)
  }
  if (!s.includes('.')) s += '.0'
  return s
}

function normalizeExponent(s: string): string {
  const m = /^(-?[0-9]*\.?[0-9]*)[eE]([+-]?[0-9]+)$/.exec(s)
  if (!m) return s
  const mant = m[1]!
  let exp = m[2]!
  if (exp.startsWith('+')) exp = exp.slice(1)
  let sign = ''
  if (exp.startsWith('-')) {
    sign = '-'
    exp = exp.slice(1)
  }
  exp = exp.replace(/^0+/, '') || '0'
  return `${mant}e${sign}${exp}`
}

function expandToPlain(f: number): string {
  const s = f.toFixed(20)
  return s.replace(/(\.[0-9]*?)0+$/, '$1').replace(/\.$/, '')
}
